Panorama Lookup commands

adambaumeister opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->

## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Description
Adds two new commands:
**panorama-lookup-route**
Searches the routing table on a firewall for the given IP address and returns the route details.
**panorama-lookup-zone**
Compares both the interface and routing table to make a determination of zone in which an IP address resides

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Demisto
- [ ] 5.0.0
- [ ] 5.5.0
- [X] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [X] No

## Must have
- [X] Tests
- [X] Documentation

adi88d commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
+
+    response = panorama_get_routes(virtual_router)
+    if 'entry' not in response['response']['result']:
+        raise Exception("No routes returned from the Firewall.")
```
```suggestion
        raise DemistoException("No routes returned from the Firewall.")
```

adi88d commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
+    Given the provided ip address, looks up the outgoing interface and zone on the firewall.
+    """
+    if not VSYS:
+        raise Exception("The 'panorama-route-lookup' command is only relevant for a Firewall instance.")
```
```suggestion
        raise DemistoException("The 'panorama-route-lookup' command is only relevant for a Firewall instance.")
```

adi88d commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
+    """
+    dest_ip = demisto.args().get("dest_ip")
+    vr = demisto.args().get("virtual_router", None)
+    demisto.results(panorama_route_lookup(dest_ip, vr))
```
please use return_results method instead,
see this [doc](https://xsoar.pan.dev/docs/integrations/code-conventions#return_results)

adambaumeister replied:
Fixed

adi88d commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
+        })
+        return r
+    else:
+        demisto.results(f"Could not map {dest_ip} to zone.")
```
please use return_results method instead,
see this [doc](https://xsoar.pan.dev/docs/integrations/code-conventions#return_results)

adi88d commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
+    vr = demisto.args().get("virtual_router", None)
+    route = panorama_route_lookup(dest_ip, vr)
+    if not route:
+        demisto.results(f"Could find a matching route to {dest_ip}.")
```
please use return_results method instead,
see this [doc](https://xsoar.pan.dev/docs/integrations/code-conventions#return_results)

adambaumeister replied:
Fixed

adi88d commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
+                    r = {**entry, **route}
+
+    if r:
+        demisto.results({
```
please use return_results method instead,
see this [doc](https://xsoar.pan.dev/docs/integrations/code-conventions#return_results)

adi88d commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
+        raise Exception("The 'panorama-route-lookup' command is only relevant for a Firewall instance.")
+
+    response = panorama_get_routes(virtual_router)
+    if 'entry' not in response['response']['result']:
```
```suggestion
    if 'entry' not in response['response'].get('result'):
```

adi88d commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
+    if 'entry' not in response['response']['result']:
+        raise Exception("No routes returned from the Firewall.")
+    else:
+        routes = response['response']['result']['entry']
```
```suggestion
        if response.get('response') and response['response'].get('result') and response['response']['result'].get('entry'):
                routes = response['response']['result']['entry']
         else:
                 routes = []
```

adi88d commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
+    current_match = None
+    matched_route = None
+    for route in routes:
+        subnet_raw = route['destination']
```
```suggestion
        subnet_raw = route.get('destination')
```

adi88d commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
+    if matched_route:
+        return matched_route
+    else:
+        raise Exception("Route not found.")
```
```suggestion
        raise DemistoException("Route not found.")
```

adi88d commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
+    interfaces = panorama_get_interfaces()
+
+    r = {}
+    if "ifnet" in interfaces["response"]["result"]:
```
```suggestion
    if "ifnet" in interfaces["response"].get("result"):
```

adi88d commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
+        demisto.results(f"Could find a matching route to {dest_ip}.")
+        return
+
+    interface = route["interface"]
```
```suggestion
    interface = route.get("interface")
```

adi88d commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -4846,6 +4849,131 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
+    r = {}
+    if "ifnet" in interfaces["response"]["result"]:
+        for entry in interfaces["response"]["result"]["ifnet"]["entry"]:
+            if entry["name"] == interface:
```
```suggestion
            if entry.get("name") == interface:
```

adi88d reviewed (changes requested):
Thanks for your contribution @adambaumeister!  it Looks great.
few notes for both `panorama-lookup-route` and `panorama-lookup-zone` commands:
- please return a human readable table containing all the output of the command + the raw response from the api, use `return_results` for this see the docs [here](https://xsoar.pan.dev/docs/integrations/code-conventions#return_results).
- Update the outputs of the commands, you can use `json-to-outputs` command in demisto-sdk for this.

adambaumeister commented:
I updated a couple of things here
* As requested, changed all teh demisto.results() calls to return_results instead (Way easier anyway!!!)
* Changed the route lookup logic to be based on the _test routing fib-lookup_ command instead - I realized that in a firewall with a lot of different paths the previous command may have returned misleading results
* Added outputs as requested and also added descriptions.
* Added some more, better tests

adi88d commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -4846,6 +4848,138 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
     )
 
 
+def panorama_get_routes(virtual_router=None):
```
remove this method if you dont use it

adambaumeister replied:
done!

adi88d commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -4846,6 +4848,138 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
+    if r:
+        # add the given dest_ip as another item int the object for simplicity
+        r["dest_ip"] = dest_ip
+        markdown = tableToMarkdown('Zone Lookup Results', r, headers=["dest_ip", "zone", "interface"])
```
please add the rest of the headers name, ip, id, etc.

adambaumeister replied:
Resolved by removing the header spec altogether and letting tableToMarkdown figure it out (there are a number of headers that may or may not appear depending on the FW config, so I figured this was the most suitable way)

adi88d commented on Packs/PAN-OS/Integrations/Panorama/Panorama.py:
```diff
@@ -4846,6 +4848,138 @@ def panorama_override_vulnerability(threatid: str, vulnerability_profile: str, d
+    r = panorama_route_lookup(dest_ip, vr)
+
+    r["dest_ip"] = dest_ip
+    markdown = tableToMarkdown('Route Lookup Results', r, headers=["dest_ip", "interface"])
```
please add the rest of the headers ip, src, metric, etc.

adi88d commented on Packs/PAN-OS/Integrations/Panorama/Panorama.yml:
```diff
@@ -4370,7 +4372,110 @@ script:
+      description: 'Forwarding Virtual Router'
+      type: String
+    - contextPath: Panorama.ZoneLookup.dyn-addr
+      description: ''
```
add description

adi88d commented on Packs/PAN-OS/Integrations/Panorama/Panorama.yml:
```diff
@@ -4370,7 +4372,110 @@ script:
+      description: 'Outgoing interface IP'
+      type: String
+    - contextPath: Panorama.ZoneLookup.id
+      description: ''
```
add description

adi88d reviewed (changes requested):
Nice! see my comments in the PR, please update the README file - use `generate-docs` in the SDK and add example for the commands  `panorama-lookup-route` and `panorama-lookup-zone`.

adi88d reviewed (changes requested):
- please update the README file - use `generate-docs` in the SDK and add example for the commands  `panorama-lookup-route` and `panorama-lookup-zone`.
- update the `palo_alto_firewall_test_pb` playbook

adambaumeister commented:
Hi, 

I tried updating the docs using generate-docs but it doesn't seem to work right. It always generates an empty "command example" string and no human readable  output.  I ran with the following command:

```
demisto-sdk generate-docs -i Panorama.yml -c "panorama-lookup-zone,panorama-lookup-route" --insecure
```

```
### panorama-lookup-route
***
Lookup a given IP address in the Firewall routing table


#### Base Command

`panorama-lookup-route`
#### Input

| **Argument Name** | **Description** | **Required** |
| --- | --- | --- |
| dest_ip | Desintation IP to lookup | Required |
| virtual_router | virtual router to check against | Optional |


#### Context Output

| **Path** | **Type** | **Description** |
| --- | --- | --- |
| Panorama.RouteLookup.dest_ip | String | The IP address tested |
| Panorama.RouteLookup.dp | String | Firewall dataplane |
| Panorama.RouteLookup.interface | String | Outgoing interface |
| Panorama.RouteLookup.ip | String | The next hop IP address |
| Panorama.RouteLookup.metric | String | Route metric |
| Panorama.RouteLookup.nh | String | next hop type |
| Panorama.RouteLookup.src | String | Outgoing interface IP |


#### Command Example
``` ```

#### Human Readable Output


```

It seems like you have to specify `-e command_examples`, but this tries to run the generation for every single command in the integration and a lot of them don't work on my test instance.

Should I manually populate these values?

Cheers
Adam

adambaumeister commented:
Added an example playbook, updated the README and merged most recent changes from Upstream.

ShirleyDenkberg commented:
@adi88d  Doc review completed.
